### ISSUE TITLE
Unable to install Ruby 2.6.5 on macOS 10.15 Catalina

### ISSUE THREAD

**Issue by bforma** [rbenv/ruby-build#1361]
Hi,

I'm trying to `rbenv install 2.6.5` after upgrading from Mojave to Catalina and I am getting the following error(s):

```
rbenv install
ruby-build: using openssl from homebrew
Downloading ruby-2.6.5.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.6/ruby-2.6.5.tar.bz2
Installing ruby-2.6.5...
ruby-build: using readline from homebrew

BUILD FAILED (OS X 10.15 using ruby-build 20191004)

Inspect or clean up the working tree at /var/folders/j8/dw62n33d4gl3y2nzr_ky09400000gn/T/ruby-build.20191008094457.36804
Results logged to /var/folders/j8/dw62n33d4gl3y2nzr_ky09400000gn/T/ruby-build.20191008094457.36804.log

Last 10 log lines:
int     sigvec(int, struct sigvec *, struct sigvec *);
                                                   ^
                                                    _Nonnull
190 warnings generated.
190 warnings generated.
216 warnings generated.
190 warnings generated.
216 warnings generated.
190 warnings generated.
221 warnings generated.
```

Any clue on how to fix this?

**Comment by bforma** [comment]
This seems to be the culprit: `file.c:23:10: fatal error: 'CoreFoundation/CFString.h' file not found`

**Comment by hsbt** [comment]
I could install 2.6.5 with Catalina

```
‚ùØ rbenv install 2.6.5
ruby-build: using openssl from homebrew
Downloading ruby-2.6.5.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.6/ruby-2.6.5.tar.bz2
Installing ruby-2.6.5...
ruby-build: using readline from homebrew
Installed ruby-2.6.5 to /Users/hsbt/.local/share/rbenv/versions/2.6.5
```

Maybey, you need to invoke `xcode-select --install` again.

**Comment by bforma** [comment]
@hsbt Thanks for the tip. Tried that and the results are the same :(

**Comment by bforma** [comment]
I've set `CPATH=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include`, otherwise there it a complaint that `stdio.h` cannot be found.

**Comment by bforma** [comment]
Got it working. Apparently I had an old version (6) of `llvm` linked via `brew` (thus also using an old version of `clang`). `brew unlink llvm@6` fixed that. Also needed to `unset CPATH`.

**Comment by mamantoha** [comment]
Thanks @bforma  . I have the same issue.

Actually unlinking `llvm` didn't help me.
I completely uninstall all `llvm` versions via `brew`. And then it was fixed.

**Comment by imbsky** [comment]
@bforma @mamantoha Try this `SDKROOT="/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk"`.

**Comment by imbsky** [comment]
I don't think this is a perfect solution, but it works well as a temporary solution.

**Comment by joshuapinter** [comment]
I had a different error but trying to do the same thing and related to `openssl`.

The system version of `openssl` was using `libressl` and an old one at that. To fix it I had to do the following: 

1. Check your existing version of `openssl`:

    ```bash
    $ openssl version
    ```

2. Backup your existing `openssl`:

    ```bash
    $ sudo mv /usr/bin/openssl /usr/bin/openssl-old
    ```

    _NOTE: You may need to turn off SIP (System Integrity Protection) by booting into Recovery Mode and running `csrutil disable`. You may also need to re-mount your root drive as writeable with `sudo mount -uw /`._

3. Install `openssl` from Brew:

    ```bash
    $ brew install openssl
    ```

4. Verify the new version:

    ```bash
    $ openssl version
    OpenSSL 1.0.2t  10 Sep 2019
    ```

**Comment by imbsky** [comment]
Any solution that requires disabling SIP is a very bad idea. and probably you just need to add the path to the OpenSSL which installed by brew.

**Comment by imbsky** [comment]
Bash:

```bash
export PATH=$PATH:/usr/local/opt/openssl/bin
```

Fish:

```fish
set -x PATH /usr/local/opt/openssl/bin $PATH
```

**Comment by gnodnooh** [comment]
I have the same issue, and I found my case was due to [conflictions with anaconda environment](https://stackoverflow.com/a/54619340/10164193). I simply deactivated conda environment of my terminal and re-installed rbenv successfully.

**Comment by joshuapinter** [comment]
> Any solution that requires disabling SIP is a very bad idea. and probably you just need to add the path to the OpenSSL which installed by brew.

To be clear, you can reenable it after updating `openssl` by rebooting into recovery mode and using `csrutil enable` in the Terminal.

> `export PATH=$PATH:/usr/local/opt/openssl/bin`

Do you know if this will use brew's openssl if the system has openssl already?

**Comment by imbsky** [comment]
Oh, probably `RUBY_CONFIGURE_OPTS` is good for ruby-build users.

```bash
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl)" rbenv install [version]
```

**Comment by imbsky** [comment]
@joshuapinter Could you try this?

**Comment by imbsky** [comment]
@bforma @mamantoha Problems on the macOS related with LLVM have been fixed in https://github.com/Homebrew/homebrew-core/pull/45304 and it has been merged today so it's already available in the latest brew.

Could you try this?

```
brew update
brew upgrade
brew uninstall llvm
brew install llvm
```

You can check statuses of other LLVM versions can be seen in https://github.com/Homebrew/homebrew-core/issues/45061#issuecomment-547671808.

**Comment by joshuapinter** [comment]
> Oh, probably `RUBY_CONFIGURE_OPTS` is good for ruby-build users.
> 
> ```shell
> RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl)" rbenv install [version]
> ```

I believe this will work as well if you're unable or unwilling to disable SIP. Give me a bit and I'll restore the old version and see if this solves it.

**Comment by cocoonkid** [comment]
I tried all of these but cannot get it work.

openssl version
OpenSSL 1.1.1d  10 Sep 2019

`rbenv install 2.6.5
Downloading openssl-1.1.1d.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/1e3a91bc1f9dfce01af26026f856e064eab4c8ee0a8f457b5ae30b40b8b711f2
Installing openssl-1.1.1d...

BUILD FAILED (OS X 10.15.1 using ruby-build 20191205)

Inspect or clean up the working tree at /var/folders/_v/q9b0p1vn60j7r231n0qfwxtc0000gn/T/ruby-build.20191207124022.48565.zs8yAo
Results logged to /var/folders/_v/q9b0p1vn60j7r231n0qfwxtc0000gn/T/ruby-build.20191207124022.48565.log

Last 10 log lines:
clang  -I. -Iinclude -fPIC -arch x86_64 -fsanitize=signed-integer-overflow -fsanitize=undefined -ggdb3 -O0 -std=c11 -Wall -Werror -Wextra -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wshadow -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/opt/rbenv/versions/2.6.5/openssl/ssl\"" -DENGINESDIR="\"/usr/local/opt/rbenv/versions/2.6.5/openssl/lib/engines-1.1\"" -D_REENTRANT -DZLIB -DZLIB_SHARED -DNDEBUG -I/usr/local/opt/rbenv/versions/2.6.5/include -I/usr/local/opt/openssl@1.1/include -MMD -MF apps/app_rand.d.tmp -MT apps/app_rand.o -c -o apps/app_rand.o apps/app_rand.c
clang  -I. -Iinclude -fPIC -arch x86_64 -fsanitize=signed-integer-overflow -fsanitize=undefined -ggdb3 -O0 -std=c11 -Wall -Werror -Wextra -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wshadow -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/opt/rbenv/versions/2.6.5/openssl/ssl\"" -DENGINESDIR="\"/usr/local/opt/rbenv/versions/2.6.5/openssl/lib/engines-1.1\"" -D_REENTRANT -DZLIB -DZLIB_SHARED -DNDEBUG -I/usr/local/opt/rbenv/versions/2.6.5/include -I/usr/local/opt/openssl@1.1/include -MMD -MF apps/apps.d.tmp -MT apps/apps.o -c -o apps/apps.o apps/apps.c
clang  -I. -Iinclude -fPIC -arch x86_64 -fsanitize=signed-integer-overflow -fsanitize=undefined -ggdb3 -O0 -std=c11 -Wall -Werror -Wextra -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wshadow -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/opt/rbenv/versions/2.6.5/openssl/ssl\"" -DENGINESDIR="\"/usr/local/opt/rbenv/versions/2.6.5/openssl/lib/engines-1.1\"" -D_REENTRANT -DZLIB -DZLIB_SHARED -DNDEBUG -I/usr/local/opt/rbenv/versions/2.6.5/include -I/usr/local/opt/openssl@1.1/include -MMD -MF apps/bf_prefix.d.tmp -MT apps/bf_prefix.o -c -o apps/bf_prefix.o apps/bf_prefix.c
clang  -I. -Iinclude -fPIC -arch x86_64 -fsanitize=signed-integer-overflow -fsanitize=undefined -ggdb3 -O0 -std=c11 -Wall -Werror -Wextra -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wshadow -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/opt/rbenv/versions/2.6.5/openssl/ssl\"" -DENGINESDIR="\"/usr/local/opt/rbenv/versions/2.6.5/openssl/lib/engines-1.1\"" -D_REENTRANT -DZLIB -DZLIB_SHARED -DNDEBUG -I/usr/local/opt/rbenv/versions/2.6.5/include -I/usr/local/opt/openssl@1.1/include -MMD -MF apps/opt.d.tmp -MT apps/opt.o -c -o apps/opt.o apps/opt.c
apps/opt.c:175:10: error: missing field 'retval' initializer [-Werror,-Wmissing-field-initializers]
    {NULL}
         ^
1 error generated.
make[1]: *** [apps/opt.o] Error 1
make: *** [all] Error 2`

**Comment by imbsky** [comment]
@cocoonkid Have you set `RUBY_CONFIGURE_OPTS`? Have you installed openssl with brew?

**Comment by jarod-legault** [comment]
> I tried but couldn't reproduce the initial issue. 🤷‍♂
> 
> So, for others, it's worth trying this solution before replacing the system's `openssl`.
> 
> 👍

@joshuapinter and @imbsky I was able to install ruby using `RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl)" rbenv install` and it worked great. Thanks for finding solutions for this issue!

**Comment by imbsky** [comment]
@jarod-legault I'm glad to hear that! Have a good rbenv life.

**Comment by joshuapinter** [comment]
> > I tried but couldn't reproduce the initial issue. 🤷‍♂
> > So, for others, it's worth trying this solution before replacing the system's `openssl`.
> > 👍
> 
> @joshuapinter and @imbsky I was able to install ruby using `RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl)" rbenv install` and it worked great. Thanks for finding solutions for this issue!

🙌 Thanks for confirming!

**Comment by iKK001** [comment]
Does not work. I tried: `RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl)" rbenv install 2.6.5`
But I still get the same Error `BUILD FAILED (OS X 10.15.2 using ruby-build ...`)

**Comment by iKK001** [comment]
What finally helped what to recognise that the Ruby version installed needed the following permission setting:

`sudo chown -R YOURUSER:staff /Users/YOURUSER/.rbenv`

Found it here :
[miroslav's solution](https://github.com/rbenv/rbenv/issues/930#issuecomment-263015645)

**Comment by hipertracker** [comment]
Nothing works. :( I use macOS Cataline 10.15.2. I just can't install Ruby from rbenv. Although `brew install ruby` works.

**Comment by imbsky** [comment]
It doesn't matter because brew only downloads and installs pre-built stuff unless you specify it with a flag.

**Comment by imbsky** [comment]
Have you tried the way I showed you above? I don't know how to help you because I don't have any information.

**Comment by iKK001** [comment]
Here is how it normally works for me (MacOS Catalina 10.15.2):

```ruby
Install other Ruby version

Use rbenv:  https://github.com/rbenv/rbenv

1. Install homebrew (https://brew.sh)
(do first line is enough)

2a. Install rbenv: Open Terminal and write:
brew install rbenv 

2b. Eventually update rbenv (if you come in later you can continue from here…):
brew upgrade rbenv ruby-build

3. Set up rbenv in your shell:
rbenv init

4. Open ~/.bash_profile   (you can write nano ~/.bash_profile)
And add the following line to your bash_profile file:
eval "$(rbenv init -)"

5. Restart your computer

7. Run doctor script - everything should look OK
curl -fsSL https://github.com/rbenv/rbenv-installer/raw/master/bin/rbenv-doctor | bash

8. Find you what Ruby version you want to install. The following list them all:
rbenv install -l

9. Install your ruby version of choice:
rbenv install 2.6.5
If Error - eventually do:
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl)" rbenv install 2.6.5
(As described here: https://github.com/rbenv/ruby-build/issues/1361)


10. Install Ruby gems:
sudo gem install bundler

11. Check the location where gems are located:
sudo gem env home
```

**Comment by hipertracker** [comment]
„Restart your compute” - really? :)

**Comment by iKK001** [comment]
smile - I used to work with Windows and this reflex is somehow in my spine... ;)
And believe it or not, the above list is what I collected from different forum entries and it helped me bullet-proof to install a new ruby version.

**Comment by hipertracker** [comment]
This is nothing new. I tried it all before. It does NOT work. Worse, I have created the latest Ubuntu server in Docker and tried to install Ruby 2.7.0 with ruby-build. It also has failed! It looks bad. 
I can install 2.6.5 with Homebrew, but I want 2.7.0.
I can install TruffleRuby from GraalVM but it can also install the older Ruby version. 
Does nobody provide binaries for Ruby 2.7.0? Strange. It has been published a month ago.

**Comment by kingshalaby1** [comment]
Remember to `rbenv rehash` after installing a new ruby version

**Comment by novibonjovi** [comment]
> > Oh, probably `RUBY_CONFIGURE_OPTS` is good for ruby-build users.
> > ```shell
> > RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl)" rbenv install [version]
> > ```
> 
> I believe this will work as well if you're unable or unwilling to disable SIP. Give me a bit and I'll restore the old version and see if this solves it.

This one worked for me.
My openssl was linked to libressl.

**Comment by henrynguyen6677** [comment]
> @bforma @mamantoha Problems on the macOS related with LLVM have been fixed in [Homebrew/homebrew-core#45304](https://github.com/Homebrew/homebrew-core/pull/45304) and it has been merged today so it's already available in the latest brew.
> 
> Could you try this?
> 
> ```
> brew update
> brew upgrade
> brew uninstall llvm
> brew install llvm
> ```
> 
> You can check statuses of other LLVM versions can be seen in [Homebrew/homebrew-core#45061 (comment)](https://github.com/Homebrew/homebrew-core/issues/45061#issuecomment-547671808).

Thank you very much.
I install rvm, rbenv and make source ruby but fired errors.
This is exactly solution for me.
```
brew update
brew upgrade
brew install llvm
rbenv install 2.7.1
rbenv rehash
```

**Comment by milushov** [comment]
Guys, does anyone experienced that ruby started load 100% CPU after upgrading macOS Catalina up to 10.15.5 (19F96)?

```
ruby -v
ruby 2.6.3p62 (2019-04-16 revision 67580) [universal.x86_64-darwin19]
```

![image](https://user-images.githubusercontent.com/424326/83506465-e8414a80-a4cf-11ea-8e30-4aee28f36b9c.png)
